### ISSUE TITLE
Grainreference - CustomStreamProvider Delay issue

### ISSUE THREAD

**Issue by angshuman-agarwal** [dotnet/orleans#981]
Hello,

I have a custom persistent stream provider and a grain implementing IAsyncObserver<int>

Inside one of the observer grain's methods, I'm subscribing to a stream that's using my custom provider:

            await stream.SubscribeAsync(this, null);

This works fine. However, if I pass the grain as a reference:
           
            // passing grain reference
            var reference = this.AsReference<ITestObserver>(); 
            await stream.SubscribeAsync(reference, null); // causes massive delays

This causes a massive delay for every message sent on the stream. Every messages arrives after about 4 seconds.

Inspecting Orleans source, I can see that the delay happens in "InvokeMethodAsync" in GrainReference.cs when it tries to invoke the remote method:

            resultTask = OrleansTaskExtentions.ConvertTaskViaTcs(resultTask);
            return (T) await resultTask;

The remote method in this case is the OnNextAsync implemented in the observer grain.

Some details:
- My custom stream provider is implementing PersistentStreamProvider
- Using a normal C# "Queue" internally

If you need the source code for the custom stream provider, please let me know, I cannot upload it now, but I can do it later.

My questions are:

- When to use AsReference?
- Why does using AsReference cause such a massive delay?

Any help is appreciated, thanks a lot

**Comment by sergeybykov** [comment]
>var reference = this.AsReference<ITestObserver>(); 
await stream.SubscribeAsync(reference, null); // causes massive delays

This is an unsupported case. The first argument to `SubscribeAsync` needs to be a local object that implements `IAsyncObeserver<T>` within the grain of the grain itself (if it implements it).

By passing a grain reference instead of a local object you are telling it to make a remote call to the grain itself, which makes no sense in this case.

**Comment by jason-bragg** [comment]
If I'm understanding what you're doing correctly, I'm surprised it works at all. :?

The stream infrastructure calls the grain via a grain extension, which is a nifty way to extend a grains interface at run time.  This means that every event being sent to your grain is already being delivered to your grain in a grain call.  By subscribing to the stream as a grain reference of yourself, I think you're causing the on next to act like a grain call to yourself.  Since you're doing this inside a grain call (made from the stream infrastructure to a grain extension attached to your grain), assuming your grain is non-reentrant, the OnNextAsyn call will not be processed until the stream call to the grain extension times out.

Basically, I think you're making a grain call to yourself in a grain call.  The grain call to yourself can't be processed until the grain call you're already in, which is awaiting the grain call to yourself, times out.

What is your grain timeout configured at?  I'm guessing ~4 seconds.. :)

**Comment by gabikliot** [comment]
Agree with Jason: "I think you're making a grain call to yourself in a grain call."
The API does not work this way, as Sergey said. 

Maybe we should add a check inside `SubsribAsync` that not grain reference as passed and throw ArgumentException.

**Comment by angshuman-agarwal** [comment]
The reason I used AsReference in the first place is because line 257 in PersistentStreamPullingAgent was timing out. The line is:

                    requestedToken = await AsyncExecutorWithRetries.ExecuteWithRetries(
                         i => consumerData.StreamConsumer.GetSequenceToken(consumerData.SubscriptionId),
                         AsyncExecutorWithRetries.INFINITE_RETRIES,
                         (exception, i) => true,
                         config.MaxEventDeliveryTime,
                         DefaultBackoffProvider);

When I used "AsReference" instead, it passed fine. Obviously, there must have been something else that was going wrong at that time.

I think there should be an error/warning given during runtime to point this out.

**Comment by amamh** [comment]
One more question (not about streams), please:

In the case of grain "A" implementing IGrainObserver, when I want to send it (from within) to another grain "B" that is doing some observable updates. I do:

    grainB.SubscribeMethod(this.AsReference<IMyGrainObserverInterface>())


Why is that different?


Ref:
http://dotnet.github.io/orleans/Getting-Started-With-Orleans/Observers

**Comment by gabikliot** [comment]
@amamh, yuou don't need to use observers for grains. Just use grain references and use grain method to notify. Observers are for clients.
In any case, you need to pass the logical actor reference. It can either be the grain key (long/Guid/string) or grain reference. It cannot be anything else.

**Comment by jason-bragg** [comment]
>"line 257 in PersistentStreamPullingAgent was timing out ... I think there should be an error/warning given during runtime to point this out."

Whenever the agent fails to deliver an event to the consumer (as would be the case if the call timed out) it makes a best effort attempt to notify the consumer of this error by calling the consumers OnErrorAsync.  However, If the OnErrorAsync call fails (times out?) the grain will have no notification of the error.

Since notifying the grain is not reliable, a more reliable means of error handling is provided by the IStreamFailureHandler.  The PersistentStreamPullingAgent is initialized with an instance of IStreamFailureHandler which is called when the agent fails to setup a subscription or fails to deliver events.  You can introduce your own implementation of IStreamFailureHandler via the IQueueAdapterFactory.

>"I have a custom persistent stream provider"

The PersistentStreamPullingAgent is part of the Persistent Stream Provider.  Is your stream provider a custom provider based off the Persistent Stream Provider or is it a Persistent Stream Provider with your own queue adapter?

**Comment by angshuman-agarwal** [comment]
Persistent Stream Provider has my own queue adapter.

**Comment by jason-bragg** [comment]
Understood.  With that in mind, I suggest your adapter provide a custom implementation of IStreamFailureHandler that can help you track down stream processing errors.

I also suggest taking a look at some of the stream processing tests (test category "Streaming") to see various working examples of simple stream processing.  The disparity between those tests and your code may help narrow what is causing the timeouts.
